### PR TITLE
Media: Check cross-origin isolation requirement for video optimization

### LINKED ISSUE #6776: Video Optimization: SharedArrayBuffer support
## Bug Description
When dragging video into canvas ( when video transcoding is enabled ), the video transcodes, however the video element fails to insert and generates a snackbar message. 

## Expected Behaviour

Video element should insert correctly. 
## Steps to Reproduce

1. Enable video transcoding.
2. Open editor in safari or firefox. 
3. Drag video into canvas. 

## Screenshots

https://user-images.githubusercontent.com/237508/111199458-8cf0cc80-85b8-11eb-9300-87a35fd3ca4b.mov

https://user-images.githubusercontent.com/237508/111199477-911cea00-85b8-11eb-8f4d-1d9b750e2a19.mov

## Additional Context

- Plugin Version:
- WordPress Version: 5.7
- Operating System: 10.15.07
- Browser: Firefox 86 / safari 14.0.1

---

_Do not alter or remove anything below. The following sections will be managed by moderators only._

### PR DESCRIPTION
**Pull request by spacedmonkey**
## Summary

Ensure the correct header are sent to browser and check if the browser supports them. 

Safari does not support transcoding at this time, failure is expected. 

### QA

- [ ] This is a non-user-facing change and requires no QA

This PR can be tested by following these steps:

### Successful transcoding ( Chrome / Firefox )
1. Enable video transcoding.
1. Open editor in firefox / chrome.
1. Drag .mov video file into canvas.
1. Video to start to transcoding. 

### Fail transcoding ( Safari )
1. Enable video transcoding.
1. Open editor in safari.
1. Drag .mov video file into canvas.
1. File should be rejected.
### UAT

- [ ] UAT should use the same steps as above.

This PR can be tested by following these steps:

1.

## Checklist

- [x] This PR addresses an existing issue and I have linked this PR to it in ZenHub
- [x] I have tested this code to the best of my abilities
- [x] I have verified accessibility to the best of my abilities ([docs](https://github.com/google/web-stories-wp/blob/main/docs/accessibility-testiing.md))
- [x] I have verified i18n and l10n (translation, right-to-left layout) to the best of my abilities
- [x] This PR contains automated tests (unit, integration, and/or e2e) to verify the code works as intended ([docs](https://github.com/google/web-stories-wp/tree/main/docs#testing))
- [x] I have added documentation where necessary
- [x] I have added a matching `Type: XYZ` label to the PR

---

Fixes #6776

### CODE REVIEW

File: assets/src/edit-story/app/media/utils/useTranscodeVideo.js [hunk 1]
```diff
@@ -124,9 +124,12 @@ function useTranscodeVideo() {
     currentUser.meta?.web_stories_media_optimization
   );
 
+  const ffmpegSupported = Boolean(window?.crossOriginIsolated);
```
Comment (swissspidy) [inline assets/src/edit-story/app/media/utils/useTranscodeVideo.js]: Let's just move this to `isFeatureEnabled` please.

Can rename it to `isFeatureSupported` if we want, but I don't want to introduce a separate variable for it.

That also means no change is needed to `uuseUploadMedia`

File: includes/Story_Post_Type.php [hunk 2]
```diff
@@ -754,4 +755,26 @@ public function filter_list_of_allowed_filetypes( $value ) {
 
 		return $value;
 	}
+
+	/**
+	 * Add cross origin headers, to make sure the ffmpeg loads.
```
Comment (swissspidy) [inline includes/Story_Post_Type.php]:
```suggestion
	 * Send headers to enable cross-origin isolation.
	 *
	 * This is required for WebAssembly threads as used for the
	 * video optimization feature powered by fffmpeg.wasm.
	 *
	 * @link https://web.dev/coop-coep/
	 * @link https://ffmpegwasm.github.io/
```

File: assets/src/edit-story/app/media/utils/useTranscodeVideo.js (hunk 1 above)
Comment (spacedmonkey) [reply assets/src/edit-story/app/media/utils/useTranscodeVideo.js]: Fixed

**Comment by spacedmonkey** [comment]
@swissspidy Product question, if we know that trancoding is not support by the browser, should we not show the first time dialog for opting in transcoding. It is a pretty weird for a user to open a browser, see a mesasge about transcoding that is just work....

CC @o-fernandez

**Review by swissspidy — APPROVED** [review]

**Comment by swissspidy** [comment]
> @swissspidy Product question, if we know that trancoding is not support by the browser, should we not show the first time dialog for opting in transcoding. It is a pretty weird for a user to open a browser, see a mesasge about transcoding that is just work....

That's already the case? 🤔  The dialog is only shown when transcoding is in progress and transcoding can only be in progress when it's supported.

**Comment by swissspidy** [comment]
Hmm the E2E tests are failing now, Puppeteer doesn't seem to like these headers 🤔 

We probably need to update Puppeteer so we're on a newer Chromium. That's a task for itself.

In the meantime, maybe passing `--disable-web-security` in `jest-puppeteer.config.cjs` helps 🤷

**Comment by spacedmonkey** [comment]
> That's already the case? 🤔 The dialog is only shown when transcoding is in progress and transcoding can only be in progress when it's supported.

It wasn't in a earlier version, but it seems to be fine now. Forget me.

**Comment by spacedmonkey** [comment]
Added QA testing details

**Comment by spacedmonkey** [comment]
As noted on Slack, I am going to share again here for visuality. 

To get a browser feature called `SharedArrayBuffers`, the origin server need to send some headers. ( required in firefox and soon to be required in Chrome in May 2021 - see blog [post](https://blog.chromium.org/2021/02/restriction-on-sharedarraybuffers.html) ).
```
Cross-Origin-Embedder-Policy: require-corp
Cross-Origin-Opener-Policy: same-origin
```

These headers limit embed resource such a script, iframes and imgs from loading only from the origin server. Meaning an resource loaded from another server will not work. See for example this avatar image. 

Screenshot 2021-03-16 at 16 17 52

To get this resources working, is a matter of adding `crossorigin` attribute in the HTML tag. More details can be found in this blog [post](https://web.dev/cross-origin-isolation-guide/#mitigate-the-impact-of-cross-origin-isolation). However, adding this attribute in places, may not be possible. There are lots of tags on the page that may load from a different domain, it maybe become a game of whack a mole to find them all. 

For this reason, I believe this feature may need to remain behind feature flag until we can resolve these issues ( if we even can ).

**Comment by spacedmonkey** [comment]
@swissspidy Regarding this PR. I recommend removing the headers added in PHP and keeping the JS. Adding a check for `crossOriginIsolated` is still valid and allow the editor to fallback to the current video uploading process, in unsupported browsers like firefox and safari. This is the expected behaviour. Thoughts?

**Comment by swissspidy** [comment]
> I recommend removing the headers added in PHP and keeping the JS. Adding a check for `crossOriginIsolated` is still valid and allow the editor to fallback to the current video uploading process, in unsupported browsers like firefox and safari. This is the expected behaviour. Thoughts?

Let's put this PR on hold for now while we consider our options in #6776.

The `crossOriginIsolated` check would still be valid, yes, but it will always be **false**, which means video optimization will never work. Which means it's a bit pointless to have in the first place.

**Comment by spacedmonkey** [comment]
Once #6875 is merged, I will remove the headers in from this PR.

File: assets/src/edit-story/app/media/utils/useFFmpeg.js [hunk 3]
```diff
@@ -64,14 +64,17 @@ function useFFmpeg() {
   const {
     state: { currentUser },
   } = useCurrentUser();
+
+  const isFeatureSupported = Boolean(window?.crossOriginIsolated);
   /**
    * Whether the video optimization feature is enabled.
    *
    * @todo Check for window.crossOriginIsolated
```
Comment (swissspidy) [inline assets/src/edit-story/app/media/utils/useFFmpeg.js]: Need to remove this note.

File: jest-puppeteer.config.cjs [hunk 4]
```diff
@@ -30,6 +30,6 @@ module.exports = {
     headless: PUPPETEER_HEADLESS !== 'false',
     slowMo: Number(PUPPETEER_SLOWMO) || 0,
     product: PUPPETEER_PRODUCT,
-    args: ['--window-size=1600,1000'], // Same as in percy.config.yml.
+    args: ['--disable-web-security', '--window-size=1600,1000'], // Same as in percy.config.yml.
```
Comment (swissspidy) [inline jest-puppeteer.config.cjs]: Can undo this I think.

File: assets/src/edit-story/app/media/utils/useFFmpeg.js (hunk 3 above)
Comment (spacedmonkey) [inline assets/src/edit-story/app/media/utils/useFFmpeg.js]:
```suggestion
```

File: jest-puppeteer.config.cjs (hunk 4 above)
Comment (spacedmonkey) [inline jest-puppeteer.config.cjs]:
```suggestion
    args: ['--window-size=1600,1000'], // Same as in percy.config.yml.
```

**Review by swissspidy — APPROVED** [review]

**Pull request merged**